wsmux/session.go

```diff
@@ -0,0 +1,221 @@
+package wsmux
+
+import (
+	"encoding/binary"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"log"
+	"net"
+	"os"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/gorilla/websocket"
+	"github.com/taskcluster/slugid-go/slugid"
+)
+
+const (
+	defaultQueueSize       = 10
+	defaultStreamQueueSize = 10
+)
+
+var (
+	errRemoteClosed  = fmt.Errorf("remote session is no longer accepting connections")
+	errAcceptTimeout = fmt.Errorf("accept timed out")
+	errBrokenPipe    = fmt.Errorf("broken pipe")
+)
+
+// Session ...
+type Session struct {
+	streamLock sync.RWMutex
+
+	streams    map[uint32]*Stream
+	writes     chan frame
+	streamChan chan *Stream
+
+	conn     *websocket.Conn
+	connLock sync.Mutex
+
+	remoteClosed bool
+	closed       bool
+	nextID       uint32
+
+	onRemoteClose  func(*Session)
+	acceptDeadline time.Duration
+	logger         *log.Logger
+}
+
+func newSession(conn *websocket.Conn, server uint32, onRemoteClose func(*Session)) *Session {
```

Comment:
Also I suggest you define an interface:
```go
type Logger interface{
  Printf(format string, a ...interface{})
}
```

And then take `Logger` as an argument... the builtin logging library can be passed as logger...
`logrus` can be used as logger, and the monitor concept we use in taskcluster-worker can be passed in...
Also please support passing in `nil`, IMO I would do that by saying that if it's nil we set it to some dummy logger that does nothing...

```go

type nilLogger struct{}
func (nilLogger) Printf(string, ...interface{}) {}

func newSession(...., log Logger) {
  if log == nil {
    log = nilLogger{}
  }
}

```

This way we always have a logger...